sematic/api/endpoints/storage.py

```diff
@@ -89,3 +93,17 @@ def get_stored_data_redirect(user: Optional[User], namespace: str, key: str):
     # response.headers.set("Cache-Control", "max-age=31536000, immutable, private")
 
     return response
+
+
+def _get_storage_destination_url(
```

Comment:
Should this logic only belong to the `LocalStorage` plugin?

Reply:
Yes, but the `AbstractStorage` interface does not allow passing a `flask.Request` object, or `origin`. We could relax the interface to allow additional args, but I don't like it.
Plus I would rather any manipulation of flask object be confined to endpoint modules.

Reply:
Can you help point out how `S3Storage` will not get into this code path?